render-wasm/src/render.rs

```diff
@@ -337,58 +381,72 @@ impl RenderState {
         debug::render(self);
     }
 
-    // Returns a boolean indicating if the viewbox contains the rendered shapes
-    fn render_shape_tree(&mut self, root_id: &Uuid, tree: &HashMap<Uuid, Shape>) -> bool {
-        if let Some(element) = tree.get(&root_id) {
-            let mut is_complete = self.viewbox.area.contains(element.bounds());
-
-            if !root_id.is_nil() {
-                if !element.bounds().intersects(self.viewbox.area) || element.hidden() {
-                    debug::render_debug_element(self, element, false);
-                    // TODO: This means that not all the shapes are rendered so we
-                    // need to call a render_all on the zoom out.
-                    return is_complete; // TODO return is_complete or return false??
-                } else {
-                    debug::render_debug_element(self, element, true);
-                }
-            }
+    pub fn render_shape_tree(
+        &mut self,
+        tree: &HashMap<Uuid, Shape>,
+        timestamp: i32,
+    ) -> Result<(), String> {
+        if !self.render_in_progress {
+            return Ok(());
+        }
 
-            let mut paint = skia::Paint::default();
-            paint.set_blend_mode(element.blend_mode().into());
-            paint.set_alpha_f(element.opacity());
-            let filter = element.image_filter(self.viewbox.zoom * self.options.dpr());
-            if let Some(image_filter) = filter {
-                paint.set_image_filter(image_filter);
-            }
+        while let Some((node_id, visited_children)) = self.pending_stack.pop() {
+            if let Some(element) = tree.get(&node_id) {
```

Comment:
This function has lots of nesting and it can be hard to read. We can refactor it later, no prob, but as a suggestion for this PR, we could handle the pattern of 

```rust
if let Some(foo) = some_option {
  // do stuff
}
else {
  return Err("Lorem ipsum")
 }
 ```
 
 In a more concise manner, with [the `?` operator](https://doc.rust-lang.org/rust-by-example/std/result/question_mark.html), assuming the function returns a `Result<_, String>` (like this one):
 
 ```rust
let foo = some_option.ok_or("Lorem ipsum".to_string())?;
 ``` 
 
 The `?` does an early return if it finds an `Err` or `None`.

Reply:
Very useful :)